### ISSUE TITLE
[Bug]: Python performance much slower than expected

### ISSUE THREAD

**Issue by jiang-qian** [termux/termux-app#3689]
### Problem description

I am participating the testing of a new universal dictionary shell
https://github.com/Crissium/SilverDict
which is mostly coded purely in python. On Android it is run on termux's verion of pyhon. Everything is being run single-threaded.

I am encounter glaring performance disparity: on my phone, the identical python code is about six time slower than when it runs on my ancient desktop running debian 12. 

But other performance benchmark suggests my phone's CPU is perfectly capable of matching if not exceeding the my desktop. For example, octane browser score on google chrome, which I think run single threaded, is higher on my phone than on my desktop. The passmark score of my phone is also much higher than my desktop.

My phone is a pixel 6 pro with Google tensor chips, my desktop is i7-2600. 

I understand that chrome and native apps like passmark android's performance cannot be compared with interpreted code like python. But I am still puzzled by the huge disparity of single threaded python performance on two CPUs with comparable raw performance.

Is there a bug in termux's python implementation? Or is there some trick I overlooked that could improve python performance on termux?

### Steps to reproduce the behavior.

Install python from pkg and all the dependencies via pip. run silverdict code with large dictionaries to be parsed by pure python code.

Termux has wakelock.

### What is the expected behavior?

Comparable performance of Python code on phone (Pixel 6 pro) and on desktop(i7-2600), whose CPU are deemed comparable by many other benchmarks.

### System information

* Termux application version:0.118.0
* Android OS version:14
* Device model:Pixel 6 pro

**Comment by romanovj** [comment]
[FAQ](https://github.com/Crissium/SilverDict/wiki/android#faq)
[Why is the server not responding/very slow?](https://github.com/Crissium/SilverDict/wiki/android#why-is-the-server-not-respondingvery-slow)
Try turning off 'battery optimisation' for Termux, which interferes with the server's ability to run in the background. Here's how to do it (on a Samsung Galaxy with Android 12): Settings -> Apps -> Termux -> Battery -> Unrestricted.

**Comment by jiang-qian** [comment]
Thank you for offering to help. Unfortunately I have always made termux
unrestricted in battery setting and all the testing had been done under
unrestricted battery optimization condition. So this is not the culprit.

Best regards
Jiang

**Comment by jiang-qian** [comment]
OK with the help of the developer of SilverDict, I did a little more benchmark and found strange and inconsistent code performance of not just python but C code under termux.

In all the following tests termux are unrestricted in battery usage, the phone (Pixel 6 pro) pluged into AC and termux has wake lock and is in forground, with screne unlocked.

For example, I ran
[benchmarks.tar.gz](https://github.com/termux/termux-app/files/13255757/benchmarks.tar.gz)

 bench.py
```python3 bench.py
```
on my desktop with i7-2600 processor, the results always hover around 4100 ms. Whereas on termux the results fluctuates wildly from run to to run, from 27000ms to 31700ms with wild outlier of 12000ms, but all much muchs slower than desktop.

with the c code I compiled the code and run it
```
clang++ -o test threading.cc -ptread
./test
```
On the desktop it is fairly consistent, around 954350, but on termux it is wildly inconsistent, going from 3768067 to 5938802 with outliers like 1680217.

Why is termux C and python code performance so wildly inconsistent?
```

**Comment by romanovj** [comment]
sdm662 cpu
1st test
```
python-speed v1.3 using python v3.11.6
multithreading: 13333.07 ms
```
2nd test is too short, (1225990 - 3451146)

**Comment by kyufie** [comment]
There are many factors that contribute to this, I think.
And it is not Termux's fault.

Please note that this is just my hypothesis.

Your phone, which runs an ARM based processor, is different than your desktop, which runs an X86 based processor.
If we're comparing just their architecture, X86 is much much more complex, in the way that they have more tools at their disposal.
For example, in X86 you can do arithmetic operation and write the result to memory in one instruction.
In ARM, you need two instructions, do the arithmetic operation and write the result in register, and then write the value in the register to memory.
Why write to the cpu register first and not to the memory directly? Because writing to the register is much easier than writing to the memory.
It makes each instruction much simpler to implement in hardware.
And that's how ARM is, their instructions are generally small and simple, you combine those simple instructions in a certain way like lego bricks to make your tools to solve problem, like for loops, if statement, etc.
X86 does this as well, it's just that their lego bricks are sometimes in the form of two or three bricks glued together, and you can even sometimes find a premade tool in there. Make things a little faster.
Another analogy is to read a word letter by letter as an ARM, as an X86, you can read the whole word at once because you have more skills at your disposal than ARM.
In summary, they are simply just two different pieces of silicon.

The other factor that I think contributes to this, is the fact that your phone's operating system is different than your desktop.
Both of them use Linux kernel, but you also need to think of what's sitting on top of it.
In Android, you have a lot of user and system processes running in the background. Sometimes, interrupting each other.
A smartphone can do more than a desktop, so you tend to need more processes for it.
Whereas Debian, not so much. What you have is what you installed.
Due to this, sometimes a process can get interrupted and temporarily replaced by another process that runs on the same CPU.
This might explain why the result is so wildly inconsistent.

The reason why the phone is not as fast as it was in the first benchmark is probably because it has slower RAM.
Even though that your phone has comparable performance against your desktop, you also need to remember that the result is taken in a controlled test that is probably designed to stress the CPU and not the RAM.
The benchmark code you provided seem to heavily make use of memory operations and only a bunch of simple arithmetic operations (Not a something that would stress the CPU that much).
Desktop RAM normally has 8 chips that can be accessed in parallel.
A phone with only 1 or 2 chips is simply no match for it.

Other factors including temperatures, the compiler being used, whether or not the device is currently servicing hardware interrupts (interaction with sensors), and even disk performance.

**Comment by romanovj** [comment]
I think it's kernel settings, I have ~13000 ms each run on slow sdm662, while author has
`27000ms to 31700ms with wild outlier of 12000ms`
on pixel 6 pro, which should be much faster

**Comment by Abdulrahman392011** [comment]
hey guys I was looking into this for a while now and I think I have a lead on what this is. 

i used ollama serve command as the test subject here.
when you use the ollama serve command in the main termux terminal you get results about 10 tokens a second.
but when you use ollama serve command in an emulated terminal inside of xfce session the performance drops to 0.07 tokens. this apply to all of the other codes and programming languages. 

I tried talking to @twaik but he said that it's not releavant to the X server. which I now believe is true. but the fact that it runs with decent performance in the main terminal inside and outside of proot and also while having a xfce session running in the background means that there is something wrong that can be fixed.

**Comment by sylirre** [comment]
@Abdulrahman392011 You have lower performance because ollama does heavy operations while CPU is being shared with other programs. All graphics under normal circumstances are not GPU accelerated and have to be computed on CPU instead.

With ADB shell and `top` utility you can see how CPU resources are distributed between processes on your device.

Termux or X11 server do not affect execution of other programs. They work independently and only process output: Termux terminal - handling text on stdout/stderr and stdin, X11 server - graphics.

This ollama issue has nothing to do with ticket where you replied. Ollama is not a Python program and regarding Python see my reply above.

**Comment by Abdulrahman392011** [comment]
Hi @sylirre . I feel that I am imposing. But genuinely, you guys keep misinterpreting what I say about this particular topic,

Again:
You run a xfce4 session, inside you open an emulated terminal and run ollama serve. Then in another emulated terminal inside of the xfce session you run ollama run "whatever" --verbose. Gives out 0.07.

Now go to the ollama serve emulated terminal press ctrl+c to stop the ollama serve then keep the xfce session running in the background and slide a new termux session from the side. Then login another proot-distro "ubuntu" type ollama serve. And slide to go back to your termux session where xfce session is running. Open up an emulated terminal and type ollama run "whatever" --verbose. It will give out 10 tokens a second.

Please guys I am not stupid, I am trying to help.

**Comment by agnostic-apollo** [comment]
>But when you use ollama serve command in an emulated terminal inside of xfce session the performance drops to 0.07 tokens. this apply to all of the other codes and programming languages.

When main Termux app is in background and X11 app is in foreground, main Termux app won't have CPU priority and any commands running in it will be throttled. This is a known issue, one of the solutions for main Termux app to show a tiny overlay too keep itself in foreground, but that hasn't been implemented yet. You can try split screen mode for now to test to know if that's the issue.

**Comment by Abdulrahman392011** [comment]
hi @agnostic-apollo . I understand what you're saying but no. just do the experiment I told you guys about and you will understand that the problem is very obvious and it's in the code, not in android not in ollama not in proot and not in xfce session using too much resources or being in the background. 

again I am sorry if I am not bringing good news but at least we now can identify and isolate the problem. next we can think of a solution for it.

**Comment by agnostic-apollo** [comment]
No idea what code you are referring to, but if you are comparing performance of native Termux processes running in app terminal, vs ones running in proot, you will obviously see a big difference.

**Comment by twaik** [comment]
Probably they mean that for some reason termux-native ollama runs much slower than proot-driven ollama despite running in same conditions (same X11 server, same mesa driver (llvmpipe), etc). May be a reason in compile flags or something.

**Comment by agnostic-apollo** [comment]
>main termux terminal you get results about 10 tokens a second.
>emulated terminal inside of xfce session the performance drops to 0.07 tokens

Native is faster according to them, as expected.

**Comment by twaik** [comment]
Oh, ok. Sorry.

**Comment by Abdulrahman392011** [comment]
> No idea what code you are referring to, but if you are comparing performance of native Termux processes running in app terminal, vs ones running in proot, you will obviously see a big difference.


both running in proot. i beg of you. run the experiment. you are refuting without running the experiment. 

I think that the issue is that when you run xserver it doesn't split the code execution into two branches. one for visual content and the other can just be refereed to a another terminal session of termux. instead it just translate everything and run it as visual content. just a hypothesis, not sure.

**Comment by twaik** [comment]
It does not work this way. As you were already told when you open Termux:X11 app the Termux app ollama works in loses focus and that causes lower CPU priority. Nothing to do with this, Termux is a regular app and can not override system behaviour.

**Comment by Abdulrahman392011** [comment]
again step by step instructions.

log in proot run xfce session open the emulated terminal run the ollama server in another emulated terminal run the model (bad performance 0.07 tokens)

leave the xfce session running but stop the ollama server, then open another termux terminal session while leaveing the xfce running and then log in proot and run the ollama server. 

go back to the xfce session and run the model in an emulated terminal and you will get (good performance 10 tokens)

**Comment by Abdulrahman392011** [comment]
this doesn't just affect ollama. any type of code you will run in an emulated terminal will be very poor performance. and it is not because of the processing power being exhausted or in the background. 

i am gonna cry, seriously. haha

**Comment by twaik** [comment]
We already know what you do. Termux app loses CPU priority when it is not in focus. End of story. I am not going to explain this again.

**Comment by Abdulrahman392011** [comment]
> We already know what you do. Termux app loses CPU priority when it is not in focus. End of story. I am not going to explain this again.

thanks for the explaination @twaik . I might misunderstood the situation. 

just saying this, I meant no offense. I was just sharing my experience.
when I code i get all sorts of issues when testing some even don't surface till the code is used by others in special use cases so I thought it might be worth it to share something I found along the way.

again thanks for your attention and time.

but for what it's worth, you haven't ran the experiment and that is a bigger issue than the bug itself. you should try working on this. everyone gets something wrong sometime, you can't think of everything.

**Comment by twaik** [comment]
As a workaround you might want to run the `termux-float` window as an overlay which will keep termux app active and "foreground".

**Comment by Abdulrahman392011** [comment]
> As a workaround you might want to run the `termux-float` window as an overlay which will keep termux app active and "foreground".

thank you @twaik  for still talking with me, I considered this priority matter long before I started bothering you guys.

the reason this doesn't apply to this case is because it's the other way around. if you read the instructions again you will find that the part where the performance is bad, is when you run the ollama server in the foreground not the background.

when you run the ollama server in the background in a seperate termux terminal session and then leave it running and swipe and return back to the emulated terminal in the xfce session and then run the model. the model runs well because the server is managed by the other termux terminal session and the all what the xfce emulated terminal is doing is sending the request to the server.

again I understand that you built and spent a lot of time to make this app and it's free so I should be thankful. and I am. it's just difficult to see something wrong and not tell you about it after all of your effort.

run the experiment yourself this really is a bug. there is something that can be improved and it will affect the performance of the system in general. I am not wasting your time. and I am not talking from inexperience. I don't know everything but I know enough to tell you this is real.

**Comment by Grimler91** [comment]
@Abdulrahman392011 please open a separate issue, and please provide more detailed steps on how to reproduce. Personally I have no idea how to run ollama in termux or proot, or how to setup and run a model in it, so I cannot reproduce your issue. Would be even better with a smaller reproducible test case to run (rather than a full language model). 

I also cannot see any information about what hardware / what type of device you are testing on, please be sure to include that when you fill in the [bug report template](https://github.com/termux/termux-app/issues/new?template=01-bug-report.yml)

**Comment by Abdulrahman392011** [comment]
@twaik, there is something else that slipped my mind.
when I run the xfce session there are multiple errors that I see in the terminal. this can also be something that indicates that there are some compatibility issues with my phone or something like this.

if you can't reproduce the bug then that is good news. it means it's not affecting every one. 

however, I am running a samsung  note 20 ultra with android 13. so it's not likely.

@Grimler91, I will open another issue, to better track the progress on this.

**Comment by sylirre** [comment]
Again, breaking everything down:

* Termux: only acts as display and input method for terminal. It does not handle execution of programs. If there will be issues with Termux code, performance won't be affected unless megabytes of text sent via stdout. Ollama doesn't send as much text, so issues definitely not on this side.

* X11 server: handles only graphics from programs using GUI. Can't affect performance of ollama by design.

* Ollama: runs directly on device unless `ollama serve` was executed in proot. So in the first case no software can affect execution of `ollama` and in second case it affected by proot.

Additionally you are dealing with such thing as process scheduling which distributes available CPU time between processes. Ollama is very resource-heavy process and on Android possibly will be affected even by swiping up/down notification bar, not saying about switching to other app like Termux:X11.

Errors in XFCE session means nothing regarding ollama. Why so, it was explained above (XFCE and ollama are separate, independent processes).

Rather than doing your own research on the cause first you are telling us that bug is somewhere in Termux, shell or whatever else (of course not in ollama itself or device, they are good and non-bugged).
